### PR TITLE
NEW #22740 Add OpenID Connect impl

### PR DESCRIPTION
**Pull request by jeritiana**
See context, use case and implementation details in #22740

### CODE REVIEW

**Comment by hregis** [comment]
@jeritiana Thank you for this contribution, but on the other hand when do you check the existence of the user in Dolibarr?

**Comment by jeritiana** [comment]
@hregis I don't check the existence of the user inside this module (or at least I didn't think I had to) because it is already handled somewhere else downstream. From my understanding of the [authentication documentation](https://wiki.dolibarr.org/index.php?title=Authentication), the goal of the login modules is to check the validity of a couple user/password. In the case of OpenID, this is delegated to the OpenID Identity Provider.
Which brings us to the following sequence:
1. The user provides his credentials login/password to the OpenID provider
2. This module checks from the OpenID provider if the authentication went well
3. If the authentication fails: the error message is displayed (e.g. Bad login/password)
  If the authentication succeeds on the OpenID provider: this module returns the login value to the calling function.
  Later, if the user (i.e the returned login) does **not** exist in Dolibarr, the error message "Unable to find user xxx" is displayed.

**Comment by hregis** [comment]
@jeritiana yes but you don't check the existence of the user in Dolibarr !

**Comment by hregis** [comment]
@jeritiana you can check the existing openid login method https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/core/login/functions_openid.php

**Comment by jeritiana** [comment]
@hregis Okay I see what you mean. Isn't this already the purpose of [main.inc.php#L869](https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/main.inc.php#L869), leading to "User not found, connexion refused" and `ErrorCantLoadUserFromDolibarrDatabase`? i.e. check the existence of the user in Dolibarr?

Or should we also add this check inside an authentication module?

**Comment by hregis** [comment]
@jeritiana in my opinion it is better to check if the user exists and is "active" before querying OpenID. In addition it will be necessary to add compatibility with Multicompany, this I can take care of.

**Comment by hregis** [comment]
@jeritiana and why not have improved the OpenID login method already existing in Dolibarr?

**Comment by jeritiana** [comment]
> @jeritiana in my opinion it is better to check if the user exists and is "active" before querying OpenID. In addition it will be necessary to add compatibility with Multicompany, this I can take care of.

Sure enough, let's try that. I will let you know of any progression. Thanks for taking care of the Multicompany compatibility.

> @jeritiana and why not have improved the OpenID login method already existing in Dolibarr?

Because the already existing login method is fully featured for OpenID 2.0. Modifying would be a breaking change for Dolibarr and there would be impacts on users, compared to adding a new next gen OpenID Connect support.

**Comment by battosai30** [comment]
Line 49 there is an error I think : 
(GETPOSTISSET['code']) 
should be
(GETPOSTISSET('code'))

File: htdocs/core/login/functions_openid_connect.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+			'grant_type'    => 'authorization_code',
+			'client_id'     => $conf->global->MAIN_AUTHENTICATION_OIDC_CLIENT_ID,
+			'client_secret' => $conf->global->MAIN_AUTHENTICATION_OIDC_CLIENT_SECRET,
+			'code'          => $auth_code,
+						'redirect_uri'  => $conf->global->MAIN_AUTHENTICATION_OIDC_REDIRECT_URL
+		];
+
+		$token_response = getURLContent($conf->global->MAIN_AUTHENTICATION_OIDC_TOKEN_URL, 'POST', http_build_query($auth_param));
+		$token_content = json_decode($token_response['content']);
+		dol_syslog("functions_openid_connect::check_user_password_openid_connect /token=".print_r($token_response, true), LOG_DEBUG);
```
Comment (battosai30) [inline htdocs/core/login/functions_openid_connect.php:64]: $token_response = getURLContent($conf->global->MAIN_AUTHENTICATION_OIDC_TOKEN_URL, 'POST', http_build_query($auth_param),1, array(),array('http', 'https'),2);
=> allow local identity provider (in my case, I use virtual machines on the same network so 192.168.xx.xx and it's blocked)

**Review by FlorentPoinsaut — CHANGES_REQUESTED** [review]

File: htdocs/core/login/functions_openid_connect.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+		if ($token_content->access_token) {
```
Comment (FlorentPoinsaut) [inline htdocs/core/login/functions_openid_connect.php:66]:
```suggestion
		if (property_exists($token_content, 'access_token')) {
```

File: htdocs/core/login/functions_openid_connect.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+ * If test is ko, reason must be filled into $_SESSION["dol_loginmesg"]
+ *
+ * @param	string	$usertotest		Login
+ * @param	string	$passwordtotest	Password
+ * @param   int		$entitytotest   Number of instance (always 1 if module multicompany not enabled)
+ * @return	string					Login if OK, '' if KO
+ */
+function check_user_password_openid_connect($usertotest, $passwordtotest, $entitytotest)
+{
+	global $conf;
```
Comment (FlorentPoinsaut) [inline htdocs/core/login/functions_openid_connect.php:37]:
```suggestion
	global $db, $conf, $langs;
```

**Review by FlorentPoinsaut — CHANGES_REQUESTED** [review]
Check if the user exists as in the original openid method

**Comment by FlorentPoinsaut** [comment]
Hi @jeritiana 
Do you want to continue to maintain this PR or would you prefer that I create a new one?
Thank you in advance,

**Comment by jeritiana** [comment]
Hi @FlorentPoinsaut , I will have a look and integrate the suggested changes. Thanks for the ping!

**Comment by philippe-opendsi** [comment]
Hi, 
For the impatient, we have put this PR in a module available here. 
https://git.open-dsi.fr/dolibarr-extension/openidconnect
Working with us with a Lemon-LDAP
ping @eldy 
Regards

**Comment by eldy** [comment]
Thanks @jeritiana 
Yes, i confirm as you suggest in the issue ticket that parameters may come from the conf.php as alternative instead of database. But this can be done in a next step (for the moment oauth2 ids are into database...
Goal is to have all param related to security in a layer executed before the code (managing the database access) and so safe from any database readability.

**Pull request merged**